godot-cell/tests/mock.rs

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! A mock implementation of our instance-binding pattern in pure rust.
+//!
+//! Used so we can run miri on this, which we cannot when we are running in itest against Godot.
+
+use std::collections::HashMap;
+use std::error::Error;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::sync::{atomic::AtomicUsize, Mutex, OnceLock};
+
+use godot_cell::{GdCell, NonAliasingGuard};
+
+struct InstanceBinding(*mut ());
+
+unsafe impl Sync for InstanceBinding {}
+unsafe impl Send for InstanceBinding {}
+
+static INSTANCE_BINDINGS: OnceLock<Mutex<HashMap<usize, InstanceBinding>>> = OnceLock::new();
+
+struct InstanceStorage<T> {
+    cell: Pin<Box<GdCell<T>>>,
+}
+
+fn binding() -> &'static Mutex<HashMap<usize, InstanceBinding>> {
+    INSTANCE_BINDINGS.get_or_init(|| Mutex::new(HashMap::new()))
+}
+
+fn register_instance<T>(instance: T) -> usize {
+    static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+    let binding = binding();
+
+    let mut guard = binding.lock().unwrap();
+
+    let key = COUNTER.fetch_add(1, std::sync::atomic::Ordering::AcqRel);
+
+    assert!(!guard.contains_key(&key));
```

Comment:
I guess it wouldn't make a big difference, but the reason for this order is that i dont wanna increment the counter unless i actually can create a new instance, and i only know i can create a new instance once the lock has been acquired.

Though since this is just a mock test it shouldn't really matter in practice

Reply:
> but the reason for this order is that i dont wanna increment the counter unless i actually can create a new instance, and i only know i can create a new instance once the lock has been acquired.

I see. Could you document these thoughts in a brief comment? 🙂